File: examples/python/plot_kinematics.py
#!/usr/bin/env python3
"""Model Health Python SDK — Download and plot kinematics for an activity.

Walks through selecting a session (your own or a public demo session),
picking an activity, downloading its kinematics CSV and plotting selected
joint angle columns against time.

Usage:
    plot_kinematics.py [<api_key>]
"""

import io
import sys

import pandas as pd
import matplotlib.pyplot as plt
from docopt import docopt

from modelhealth import (
    ModelHealthService,
    ModelHealthError,
    ActivityStatus,
    MotionDataType,
)
from _prompts import confirm, pick_one, pick_multi
from _utils import save_file, MOTION_DATA_EXT, load_api_key

# ---------------------------------------------------------------------------
# Constants
# ---------------------------------------------------------------------------

DEMO_SESSION_ID = "63f3faaf-712d-4cef-845e-e2eb904ca5b4"

# Activities created by the mobile app for internal use — exclude from lists.
_INTERNAL_ACTIVITY_NAMES = {"calibration", "neutral"}


# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------

def main(api_key):
    # Ask whether to use own data or the demo session
    use_demo = confirm("Load data from the public demo session?", default=False)

    print("\nConnecting...")
    try:
        service = ModelHealthService(api_key)
    except ModelHealthError as exc:
        sys.exit(f"Failed to initialise: {exc}")

    # Load session
    if use_demo:
        print(f"\nFetching demo session '{DEMO_SESSION_ID}'...")
        try:
            session = service.get_session(DEMO_SESSION_ID)
        except ModelHealthError as exc:
            sys.exit(f"Failed to fetch demo session: {exc}")
    else:
        print("\nFetching sessions...")
        sessions = service.session_list()
        if not sessions:
            sys.exit(
                "No sessions found. Create a session using the Model Health mobile app first."
            )
        print(f"\n{len(sessions)} session(s):\n")
        session = pick_one(
            sessions,
            "Select session",
            lambda s: f"[session ID: {s.id}]  session name: {s.session_name or '(unnamed)'}  subject: {s.name or '(unnamed)'}",
        )

    # Load activities
    session_label = session.session_name or session.name or session.id
    print(f"\nFetching activities for session '{session_label}'...")
    all_activities = service.activity_list(session)
    activities = [a for a in all_activities if a.name not in _INTERNAL_ACTIVITY_NAMES]
    if not activities:
        sys.exit("No activities found in this session.")

    print(f"\n{len(activities)} activity/activities:\n")
    activity = pick_one(
        activities,
        "Select activity",
        lambda a: f"{a.name or a.id}  [{a.status}]",
    )

    # Check status
    activity_label = activity.name or activity.id
    print(f"\nChecking status of '{activity_label}'...")
    status = service.activity_status(activity)
    if status != ActivityStatus.ready:
        sys.exit(
            f"Activity '{activity_label}' is not ready (status: {status}). "
            "Wait for processing to complete and try again."
        )
    print("Activity is ready.")

    # Download kinematics CSV
    print("\nDownloading kinematics CSV...")
    results = service.motion_data_for_activity(activity, [MotionDataType.kinematics_csv])
    if not results:
        sys.exit("No kinematics CSV data returned for this activity.")

    csv_data = results[0].data
    ext = MOTION_DATA_EXT.get(results[0].type, "csv")
    slug = activity_label.replace(" ", "_")
    csv_path = save_file(f"{slug}_kinematics.{ext}", csv_data)
    print(f"  Saved: {csv_path}")

    # Load CSV
    df = pd.read_csv(io.BytesIO(csv_data), comment="#")

    # Identify the time column (first column) and data columns (rest)
    time_col = df.columns[0]
    data_cols = list(df.columns[1:])

    if not data_cols:
        sys.exit("CSV contains no data columns to plot.")

    # Let the user pick which columns to plot
    print(f"\n{len(data_cols)} column(s) available:\n")
    selected_cols = pick_multi(
        data_cols,
        "Select columns to plot",
        lambda c: c,
    )

    # Plot
    fig, ax = plt.subplots(figsize=(12, 5))
    for col in selected_cols:
        ax.plot(df[time_col], df[col], label=col)

    ax.set_xlabel(time_col)
    ax.set_ylabel("Value")
    ax.set_title(f"Kinematics — {activity_label}")
    ax.legend()
    fig.tight_layout()

    plot_path = save_file(f"{slug}_kinematics.png", b"")  # reserve filename
    fig.savefig(plot_path)
    plt.close(fig)
    print(f"\nPlot saved: {plot_path}")

    print("\nDone.")


if __name__ == "__main__":
    args = docopt(__doc__)
    main(load_api_key(args["<api_key>"]))

File: examples/ts/plot_kinematics.ts
/**
 * Model Health TypeScript SDK — download and inspect kinematics for an activity.
 * Mirrors examples/python/plot_kinematics.py.
 *
 * Saves the kinematics CSV and prints available column names. Open the CSV in
 * a spreadsheet application (Numbers, Excel, Google Sheets) or a Python/R
 * script to produce a plot.
 *
 * Usage:
 *   npx tsx plot_kinematics.ts [<api_key>]
 */

import { ModelHealthService } from '@modelhealth/modelhealth';
import {
  loadApiKey, INTERNAL_ACTIVITY_NAMES,
  pickOne, saveFile, closePrompts,
} from './_shared.js';

async function main() {
  const args = process.argv.slice(2);
  console.log('Connecting...');
  const service = new ModelHealthService({ apiKey: loadApiKey(args[0]), autoInit: false });
  await service.init();

  // Session
  console.log('\nFetching sessions...');
  const sessions = await service.sessionList();
  if (!sessions.length) {
    console.error('No sessions found. Create a session using the Model Health mobile app first.');
    process.exit(1);
  }

  console.log(`\n${sessions.length} session(s):\n`);
  const session = await pickOne(sessions, 'Select session', s => {
    const sn = s.sessionName || '(unnamed)';
    const sub = s.name || '(unnamed)';
    return `[session ID: ${s.id}]  session name: ${sn}  subject: ${sub}`;
  });

  // Activity
  const sessionLabel = session.sessionName || session.name || session.id;
  console.log(`\nFetching activities for session '${sessionLabel}'...`);
  const allActivities = await service.activityList(session.id);
  const activities = allActivities.filter(a => !INTERNAL_ACTIVITY_NAMES.has(a.name ?? ''));
  if (!activities.length) { console.error('No activities found in this session.'); process.exit(1); }

  console.log(`\n${activities.length} activity/activities:\n`);
  const activity = await pickOne(activities, 'Select activity', a => `${a.name ?? a.id}  [${a.status}]`);

  // Check status
  const activityLabel = activity.name ?? activity.id;
  console.log(`\nChecking status of '${activityLabel}'...`);
  const status = await service.activityStatus(activity);
  if (status.type !== 'ready') {
    console.error(`Activity '${activityLabel}' is not ready (status: ${status.type}). Wait for processing to complete and try again.`);
    process.exit(1);
  }
  console.log('Activity is ready.');

  // Download kinematics CSV
  console.log('\nDownloading kinematics CSV...');
  const results = await service.motionDataForActivity(activity, ['kinematics_csv']);
  if (!results.length) {
    console.error('No kinematics CSV data returned for this activity.');
    process.exit(1);
  }

  const csvData = results[0].data;
  const slug = activityLabel.replace(/ /g, '_');
  const csvPath = saveFile(`${slug}_kinematics.csv`, csvData);
  console.log(`  Saved: ${csvPath}`);

  // Parse headers
  const csvText = new TextDecoder().decode(csvData);
  const dataLines = csvText.split('\n').filter(l => l.trim() && !l.startsWith('#'));
  const headers = dataLines[0]?.split(',').map(h => h.trim()) ?? [];

  if (headers.length < 2) {
    console.log('\nCSV contains no data columns to display.');
    console.log('\nDone.');
    return;
  }

  const timeCol = headers[0];
  const dataCols = headers.slice(1);

  console.log(`\nTime column:  ${timeCol}`);
  console.log(`\n${dataCols.length} data column(s):\n`);
  dataCols.forEach((col, i) => console.log(`  ${String(i + 1).padStart(3)}. ${col}`));

  console.log(`\nThe kinematics CSV has been saved to:\n  ${csvPath}`);
  console.log('\nOpen it in Numbers, Excel, or Google Sheets to plot joint angles over time.');
  console.log('\nDone.');
}

main().catch(err => { console.error(`Error: ${err.message ?? err}`); process.exit(1); })
  .finally(closePrompts);
